`update-from-overlay.py`: A cleaner bulk-update script

AndersonTorres opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jian-lin reviewed (changes requested):
An important part of bumping elisp pkgs is to find new failed builds and fix them or report them to upstreams, which this PR seems not include.

I think it is a good idea to focus on one thing in a PR.  Let's discuss your python rewrite here and do the elisp bump in another PR.  I'll do the bump if nobody does it first.

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,102 @@
+
+# External Dependencies:
+import github
+import pygit2
```
another choice is [GitPython](https://github.com/gitpython-developers/GitPython), maybe more pythonic

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,102 @@
+
+default_branch = 'master'
+g = github.Github()
+emacsOverlayRepo = g.get_repo('nix-community/emacs-overlay')
```
```suggestion
emacs-overlay-repo = g.get_repo('nix-community/emacs-overlay')
```

jian-lin reviewed (changes requested):
This feature is very useful and easy to add, so I'll mark it as blocking.

AndersonTorres commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,102 @@
+
+if __name__ == '__main__':
+    with contextlib.chdir(heredir):
+        sha = emacsOverlayRepo.get_branch(default_branch).commit.sha
```
Nice idea! Done.

jian-lin commented:
It seems this PR is under active development. I'll unsubscribe it.  Feel free to ping me when it is ready for review.

AndersonTorres commented:
@jian-lin now that I wrote the docstrings, feel free to review.

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,202 @@
+    argv -- The argument list; default sys.argv.
+    """
+    parser = argparse.ArgumentParser(
+        prog = argv[0],
```
nitpicking:

The [default value][1] is similar and maybe is good enough?

[1]: https://docs.python.org/3/library/argparse.html#prog

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,202 @@
+    parser = argparse.ArgumentParser(
+        prog = argv[0],
+        description = 'Fetch and commit Elisp package sets from nix-community/emacs-overlay',
+        epilog = 'NixOS/Nixpkgs'
```
nitpicking:
What is the motivation for adding this line to `--help` message?  A bit confusing to see `NixOS/Nixpkgs` at the last line:

```
usage: update-from-overlay.py [-h] [--commit COMMIT]

Fetch and commit Elisp package sets from nix-community/emacs-overlay

options:
  -h, --help       show this help message and exit
  --commit COMMIT  Commit to be fetched, in SHA format. If not specified, retrieve the
                   master tip.

NixOS/Nixpkgs
```

AndersonTorres replied:
It is a bit paranoid: I was testing the functionality, and to ensure no hidden path was ignored, I filled the default parameters.

jian-lin reviewed (commented):
I skimmed through the code and it looks good.  I gave some nitpickings (Sorry!)

AndersonTorres commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,202 @@
+heredir = os.path.dirname(os.path.realpath(__file__))
+r = pygit2.Repository(heredir)
+# pygit2 uses workdir (the one with .git) as "root" to deal with files.
+relative_heredir = heredir.removeprefix(r.workdir)
```
Nice to know! I will convert it.

AndersonTorres commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,202 @@
+heredir = os.path.dirname(os.path.realpath(__file__))
+r = pygit2.Repository(heredir)
+# pygit2 uses workdir (the one with .git) as "root" to deal with files.
+relative_heredir = heredir.removeprefix(r.workdir)
```
Well, most of them are read-only magic constants.

- `elisp_packages_set` clearly describes the layout of emacs-overlay.
- `heredir` points to physical location of script (it is just the "`SCRIPT_DIR` trick" written in Python)
- `repository` is the Git handle for the repo in `heredir`, while `relative_heredir` is here merely because pygit2 does not deal with absolute file paths
- `logger` is complicated, because basically Python logging is global.

EDIT:

Now that I am reading the code, `relative_heredir` is used only by `commit_fileset`. I will move this variable inside the function.

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,203 @@
+        sha = result.stdout.split()[0]
+    else:
+        log.warning('git command not found')
+        sha = None
```
`None` will fail in the following function calls.

What about using `"master"` as the default value?

Another choice is to fail fast here.

In addition, since we use nix shebang to install `git`, maybe it is reasonable assume `git` is always available and remove this check?

AndersonTorres replied:
>  --commit should also default to "master", I think.

It would imply resolving arbitrary references, like tags, versions, branches etc. Is this suitable?

jian-lin replied:
> > --commit should also default to "master", I think.
> 
> It would imply resolving arbitrary references, like tags, versions, branches etc. Is this suitable?

I do not have a strong opinion. Feel free to ignore that specific suggestion.

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,203 @@
+    url = f'https://raw.githubusercontent.com/nix-community/emacs-overlay/{sha}/{location}/{basename}'
+
+    log.debug(f'Getting {url}')
+    tmpfile = urllib.request.urlretrieve(url)[0]
```
`urlretrieve` is [a legacy interface and might become deprecated at some point in the future][1].  It's better to use something future-proofing.

[1]: https://docs.python.org/3/library/urllib.request.html#legacy-interface

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,203 @@
+    message = f'''{attrs}: Updated at {datestring} (from nix-community/emacs-overlay)
+
+Commit: {sha}
+File: {commit_file}
```
```suggestion
```

It's not very useful, I guess.  In addition, `git log` has this info.

AndersonTorres commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,203 @@
+)
+logger = logging.getLogger('update-from-overlay')
+
+def grab_master_sha (log = logger.getChild('grab_master_sha')):
```
My intention was create a hierarchy of loggers. But now I see it is too overkill. I will rewrite this.

AndersonTorres commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,203 @@
+{message}
+''')
+        tree = repository.index.write_tree()
+        repository.create_commit(ref, author, commiter, message, tree, parents)
```
`pygit2` does not sign commits by itself, but it allows committing signed blobs. I will test this.

jian-lin replied:
FWIW, calling `git` is an alternative.

Commit:
emacs: Add update-from-overlay.py

This is a new script written in Python 3 that fetches, checks and commits the
elisp package sets from nix-community/emacs-overlay.

AndersonTorres commented:
Now the logfile is "split":
- Screen shows info messages
- Logfile shows debug and info messages

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,243 @@
+
+@dataclasses.dataclass
+class EmacsOverlay:
+    git_url = f'https://github.com/nix-community/emacs-overlay'
```
```suggestion
    git_url: str = f'https://github.com/nix-community/emacs-overlay'
```

https://docs.python.org/3/library/dataclasses.html

> The member variables to use in these generated methods are defined using [PEP 526](https://peps.python.org/pep-0526/) type annotations.

jian-lin replied:
```suggestion
    git_url: str = f'https://github.com/nix-community/emacs-overlay'
```

this applies to many places

https://docs.python.org/3/library/dataclasses.html

> The member variables to use in these generated methods are defined using [PEP 526](https://peps.python.org/pep-0526/) type annotations.

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,254 @@
+@dataclasses.dataclass
+class EmacsOverlay:
+    git_url : str = f'https://github.com/nix-community/emacs-overlay'
+    raw_url : str = f'https://raw.githubusercontent.com/nix-community/emacs-overlay'
```
```suggestion
    git_url : str = 'https://github.com/nix-community/emacs-overlay'
    raw_url : str = 'https://raw.githubusercontent.com/nix-community/emacs-overlay'
```

no need to use f-strings

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,254 @@
+    '''
+    logger.debug('BEGIN')
+
+    location = emacs_overlay.elisp_packages_set[name]['location']
```
What about replacing the `emacs_overlay` function argument with `emacs_overlay.elisp_packages_set[name]` since we only need `emacs_overlay.elisp_packages_set[name]` in this function?

This applies to many other places.

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,254 @@
+    emacs_overlay = EmacsOverlay()
+    here_directory = HereDirectory()
+    argument_parser = get_argument_parser()
+    with contextlib.chdir(here_directory.path):
```
This seems to be a no-op since `path` is just `.`.  Maybe remove `chdir`?

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,254 @@
+    def master_sha (self) -> str:
+        '''Return the SHA of current master tip.'''
+        cmdline = ['git', 'ls-remote', '--branches', self.git_url, 'refs/heads/master']
+        result = subprocess.run(cmdline, capture_output = True, text = True)
```
```suggestion
        result = subprocess.run(cmdline, capture_output = True, text = True, check=true)
```

exit if users do not provide a hash and we fail to get the latest hash

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,254 @@
+
+    args = argument_parser.parse_args()
+
+    if args.commit == None:
```
```suggestion
    if args.commit is None:
```

found by `ruff check`

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,254 @@
+    nix_attrs = emacs_overlay.elisp_packages_set[name]['nix_attrs']
+    basename = emacs_overlay.elisp_packages_set[name]['basename']
+
+    if datestring == None:
```
```suggestion
    if datestring is None:
```

found by `ruff check`

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,254 @@
+
+    for nix_attr in emacs_overlay.elisp_packages_set[name]['nix_attrs']:
+
+        cmdline = ['nix-instantiate', '--show-trace', here_directory.git_root(),
```
```suggestion
        cmdline = ['nix-instantiate', '--show-trace', str(here_directory.git_root()),
```

make mypy happy

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,254 @@
+    def git_root(self) -> pathlib.Path:
+        '''Returns the root directory of Git repository.'''
+        cmdline = ['git', 'rev-parse', '--show-toplevel']
+        result = subprocess.run(cmdline, capture_output = True, text = True)
```
```suggestion
        result = subprocess.run(cmdline, capture_output = True, text = True, check=True)
```

If this fails, then `check_fileset` fails.  It makes sense to exit if we cannot test the result (via `check_fileset`).

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,254 @@
+    console_formatter = logging.Formatter(fmt='%(asctime)s - %(levelname)s - %(funcName)s - %(message)s',
+                                          datefmt='[%Y-%m-%d %H:%M:%S]')
+
+    console_handler = logging.StreamHandler(stream = sys.stdout)
```
log usually goes to `stderr`, I think.  Do not forget to change the docstring.

I am fine with either stdout or stderr.

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,254 @@
+    parser.add_argument(
+        '--commit',
+        help = 'Commit to be fetched, in SHA format. If not specified, retrieve the master tip.',
+        default = None
```
```suggestion
```

nit: it's the default

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/update-from-overlay.py:
```diff
@@ -0,0 +1,254 @@
+
+    logger.debug('END')
+
+def commit_fileset (name: str, sha: str, datestring : str | None,
```
```suggestion
def commit_fileset (name: str, sha: str, datestring : str,
```

No not forget to also update the docstring and code of this function.

It's impossible to be `None`.

jian-lin reviewed (approved):
This script works and #494565 is created by it.  So I plan to merge it.

I also left a few suggestions.  I am willing to improve the script according to those suggestions if you are not, after the merge.
